oci-registry/build-multi-arch.sh

```diff
@@ -0,0 +1,80 @@
+#!/bin/sh
+
+#
+# Copyright Red Hat
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+buildfolder="$(basename "$(dirname "$0")")"
+# Due to command differences between podman and docker we need to separate the process
+# for creating and adding images to a multi-arch manifest
+podman=${USE_PODMAN:-false}
+# Stores all created image tags 
+images=()
+# Base Repository
+BASE_REPO="quay.io/devfile/oci-registry"
+BASE_TAG="next"
+
+function build {
+    IMAGE="$BASE_REPO:$2"
+
+    echo "Building: ${IMAGE}"
+    $1 build -t $IMAGE --platform "linux/$2" "$buildfolder"
+
+    echo "Tagging: ${IMAGE}"
+    $1 tag "$IMAGE" "$IMAGE"
+
+    echo "Pushing: ${IMAGE}"
+    $1 push "$IMAGE"
+
+    # Add image to list of all images to be added to a manifest
+    images+=("${IMAGE}")
+}
+
+function engine-handler {
+    for arch in amd64 arm64 ; do
+        build "$1" "$arch"
+    done
+}
+
+
+if [ ${podman} == true ]; then
+  echo "Executing with podman"
+
+  # Build and push multi-arch images
+  engine-handler podman
+
+  # Create manifest and add images
+  podman manifest create oci-registry-manifest
+  for img in "${images[@]}" ; do
+    podman manifest add oci-registry-manifest "$img"
+  done
+
+  # Push and delete local manifest
+  podman manifest push oci-registry-manifest "$BASE_REPO":"$BASE_TAG"
```

Comment:
QQ: We push the image having a different tag. Is this expected?

Reply:
Do you think it might be better to assign the "$BASE_REPO:$BASE_TAG" to a variable so we can use the variable instead?

Reply:
> QQ: We push the image having a different tag. Is this expected?

For multi-arch when we build for `linux/arm64` and `linux/amd64` for example we will have 3 items at the end of the "build" stage:
1. An arm64 image for `oci-registry`
2. an amd64 image for `oci-registry`
3. A manifest that has 2 pointers pointing to the arm64 image and the amd64 image

The manifest will be tagged `quay.io/devfile/oci-registry:next` for example, so when a user runs something like `docker pull quay.io/devfile/oci-registry:next` the manifest is going to allow them to receive the proper image for their architecture as it will redirect the pull to the proper image. I am on arm64 so without needing to pass anything if I run `docker pull quay.io/devfile/oci-registry:next` I will pull an arm64 image.

The arm64 and amd64 images have the tag `amd64` and `arm64` because that also enables someone to pull that particular image if they want as well.



> Do you think it might be better to assign the "$BASE_REPO:$BASE_TAG" to a variable so we can use the variable instead?

Yeah I can implement that change!

Reply:
> The manifest will be tagged quay.io/devfile/oci-registry:next for example, so when a user runs something like docker pull quay.io/devfile/oci-registry:next the manifest is going to allow them to receive the proper image for their architecture as it will redirect the pull to the proper image. I am on arm64 so without needing to pass anything if I run docker pull quay.io/devfile/oci-registry:next I will pull an arm64 image.

Nice thanks for the explanation! IMO I like this approach and I think it makes sense.